File: src/ts/lib/_line_tokenizer.ts
import {Definition, FileResponse, Token, Usage} from "./_model";
import {combineUsagesAndDefinitions, sortAndFilterTokens} from "./_helpers";

export class LineTokens {
  constructor(
    public line: number,
    public tokens: Token[]
  ) {}


  static build(fileResponse: FileResponse): LineTokens[] {
    let tokens = sortAndFilterTokens(combineUsagesAndDefinitions(fileResponse.usages, fileResponse.definitions));
    let lines = [];

    let currentLine: number | null = null;
    let currentTokens: Token[] = [];

    for (let token of tokens) {
      if (!token.location) { continue; }
      if (currentLine === null) { currentLine = token.location.start.line; }

      if (currentLine === token.location.start.line) {
        currentTokens.push(token);
      } else {
        lines.push(new LineTokens(currentLine, currentTokens));
        currentTokens = [token];
        currentLine = token.location.start.line;
      }
    }

    if (currentTokens.length > 0 && currentLine) {
      lines.push(new LineTokens(currentLine, currentTokens));
    }

    return lines;
  }
}

export class LineTokenizer {
  readonly repoName: string;
  readonly revision: string;
  readonly branch: string | null;
  readonly selectedNodeId: string | null;
  readonly tokens: Token[];
  readonly line: number;

  index: number = 0;
  col: number = 1;

  shouldBeHighlighted: boolean = false;

  constructor(options: {
    repoName: string;
    revision: string;
    branch: string | null;
    selectedNodeId: string | null;
    line: number;
    sortedTokens: Token[]
  }) {
    this.repoName = options.repoName;
    this.revision = options.revision;
    this.branch = options.branch;
    this.selectedNodeId = options.selectedNodeId;
    this.tokens = options.sortedTokens;
    this.line = options.line;
  }

  makeUrl(token: Token): string {
    switch (token.type) {
      case "usage":
        let usage = token as Usage;
        if (!usage.definition || !usage.definition.location) {
          return 'javascript:return false;';
        }

        if (usage.definition.module == 'Jdk') {
          return `http://localhost:9000/github/${this.repoName}/${this.revision}/jdk/${usage.definition.location.path}?p=${usage.definition.nodeId}`;
        } else if (usage.definition.module == 'Jar') {
          return `http://localhost:9000/github/${this.repoName}/${this.revision}/jar/${usage.definition.jarId}/${usage.definition.location.path}?p=${usage.definition.nodeId}`;
        } else if (usage.definition.module == 'User') {
          return `/${this.repoName}/blob/${this.branch || this.revision}/${usage.definition.location.path}?p=${usage.definition.nodeId}#L${usage.definition.location.start.line}`;
        } else {
          throw `Unrecognized module ${usage.definition.module}`;
        }
      case "definition":
        let definition = token as Definition;

        if (!definition.location) {
          return 'javascript:return false;';
        }

        return `http://localhost:9000/github/${this.repoName}/${this.revision}/usage/${definition.nodeId}`;

      default:
        throw `Unrecognized token type ${token.type}`;
    }
  }

  static getRelevantNodeId(token: Token): string {
    let nodeId: string = '';
    switch (token.type) {
      case "usage":
        nodeId = (token as Usage).definition.nodeId;
        break;
      case "definition":
        nodeId = (token as Definition).nodeId;
        break;
    }
    return nodeId;
  }

  private modify(elem: Node): Node[] | null {
    let s = elem.nodeValue!;
    let elemStart = this.col;
    let elemEnd = this.col + s.length - 1;

    let relevantTokens = [];

    while (
      this.index < this.tokens.length &&
      this.tokens[this.index].location!.start.line == this.line &&
      this.tokens[this.index].location!.start.col >= elemStart &&
      this.tokens[this.index].location!.start.col <= elemEnd
      ) {
      relevantTokens.push(this.tokens[this.index]);
      this.index++;
    }

    if (!relevantTokens) { return null; }

    let nodes = [];
    let start = 0;

    relevantTokens.forEach((token) => {
      let tStart = token.location!.start.col - this.col;
      let tEnd = token.location!.end.col - this.col;

      if (start > tStart) { return; }

      if (tStart > 0) {
        nodes.push(document.createTextNode(s.substring(start, tStart)));
      }

      let anchor = document.createElement('a');
      anchor.href = this.makeUrl(token);
      anchor.classList.add('codelab-link');
      if (this.selectedNodeId && this.selectedNodeId == LineTokenizer.getRelevantNodeId(token)) {
        anchor.classList.add('codelab-highlighted');
        this.shouldBeHighlighted = true;
      }
      anchor.text = s.substring(tStart, tEnd + 1);
      nodes.push(anchor);

      start = tEnd + 1;
    });

    if (start <= (s.length - 1)) {
      nodes.push(document.createTextNode(s.substring(start)));
    }

    return nodes;
  }

  private runIndex(): void {
    while (this.index < this.tokens.length && this.tokens[this.index].location!.start.col < this.col) {
      this.index++;
    }
  }

  private walk(elem: Node): Node[] | null {
    this.runIndex();

    if (elem.nodeType == Node.TEXT_NODE) {
      let size = elem.nodeValue!.length;
      let newChildren = this.modify(elem);
      this.col += size;

      return newChildren;
    } else {
      let childNodes: ChildNode[] = [];
      elem.childNodes.forEach((child) => childNodes.push(child));

      for (let child of childNodes) {
        let newChildNodes = this.walk(child);
        if (newChildNodes) {
          child.replaceWith(...newChildNodes);
        }
      }

      return null;
    }
  }

  process(elem: Node): boolean {
    this.col = 1;
    this.walk(elem);
    return this.shouldBeHighlighted;
  }
}

File: scala/content/src/main/scala/content/tokenizer/LineTokenizer.scala
package content.tokenizer

import content.bindings.Tippy
import models.{Definition, Token, Usage, UsageCount, UsageDefinition}
import org.scalajs.dom
import org.scalajs.dom.Element
import org.scalajs.dom.raw.Node

class LineTokenizer(
  val repoName: String,
  val revision: String,
  val path: String,
  val host: String,
  val branchOpt: Option[String],
  val selectedNodeIdOpt: Option[String],
  val lineTokens: LineTokens
) {

  val tokens = lineTokens.tokens
  val line = lineTokens.line

  var col: Int = 1
  var shouldLineBeHighlighted: Boolean = false

  def getUsageUrl(count: UsageCount): String = count.module match {
    case "Jdk" => s"$host/github/$repoName/$revision/jdk/${count.path}?p=${count.nodeId}"
    case "Jar" => s"$host/github/$repoName/$revision/${count.jarOpt.map(_.id).get}/${count.path}?p=${count.nodeId}"
    case "User" => s"/$repoName/blob/${branchOpt.getOrElse(revision)}/${count.path}?p=${count.nodeId}#L${count.firstLine}"
  }

  def makeUrlOpt(token: Token): Option[String] = token match {
    case usage: Usage =>
      usage.definition.locationOpt
        .map { location =>
          usage.definition.module match {
            case "Jdk" => s"$host/github/$repoName/$revision/jdk/${location.path}?p=${usage.definition.nodeId}"
            case "Jar" => s"$host/github/$repoName/$revision/jar/${usage.definition.jarOpt.get.id}/${location.path}?p=${usage.definition.nodeId}"
            case "User" => s"/$repoName/blob/${branchOpt.getOrElse(revision)}/${location.path}?p=${usage.definition.nodeId}#L${location.start.line}"
          }
        }
    case definition: Definition =>
      Some(
        if (definition.counts.isEmpty) {
          "javascript: return false;"
        } else if (definition.counts.size == 1) {
          getUsageUrl(definition.counts.head)
        } else {
          s"$host/github/$repoName/$revision/usage/${definition.nodeId}"
        }
      )
  }

  def getModuleName(definition: UsageDefinition): String = definition.module match {
    case "Jdk" => "JDK"
    case "Jar" => definition.jarOpt.map { j => s"${j.group}:${j.artifact}:${j.version}" }.get
    case "User" => repoName
  }

  def getModuleName(count: UsageCount): String = count.module match {
    case "Jdk" => "JDK"
    case "Jar" => "jars"
    case "User" => repoName
  }

  def renderOccurrenceWord(count: Int): String = {
    if (count == 1) {
      "occurrence"
    } else {
      "occurrences"
    }
  }

  def renderFileWord(count: Int): String = {
    if (count == 1) {
      "file"
    } else {
      "files"
    }
  }

  def makeTooltipContent(token: Token): String = token match {
    case u: Usage =>
      u.definition.locationOpt
        .map { location =>
          if (location.path == path && u.definition.module == "User") {
            s"Defined in this file on the line ${location.start.line}"
          } else {
            s"Defined in ${location.path} inside ${getModuleName(u.definition)}"
          }
        }
        .getOrElse(s"Defined inside ${getModuleName(u.definition)}")
    case d: Definition =>
      if (d.counts.isEmpty) {
        "No occurrences found"
      } else if (d.counts.size == 1) {
        val first = d.counts.head

        if (first.module == "User" && first.path == path) {
          s"Found ${first.count} ${renderOccurrenceWord(first.count)} only in this file"
        } else {
          s"Found ${first.count} ${renderOccurrenceWord(first.count)} only in ${first.path} inside ${getModuleName(first)}"
        }
      } else {
        val (thisFileCountOpt, otherFileCounts) = d.counts.partition { d => (d.module, d.path) == ("User", path) }

        val thisFileLabel = thisFileCountOpt
          .map { count =>
            s"""Found <a href='${getUsageUrl(count)}'>${count.count} ${renderOccurrenceWord(count.count)} in this file</a>"""
          }

        val otherFileLabels = otherFileCounts
          .groupBy { c => c.module }
          .mapValues { u => (u, u.size, u.map(_.count).sum) }
          .toList
          .sortBy(_._1)
          .map { case (_, (count, fileCount, occurrenceCount)) =>
            s"Found $occurrenceCount ${renderOccurrenceWord(occurrenceCount)} inside ${getModuleName(count.head)} ($fileCount ${renderFileWord(d.counts.size)})"
          }

        (thisFileLabel ++ otherFileLabels).mkString("<br/>")
      }
  }

  def modify(node: Node): Option[Seq[Node]] = {
    val elemStart = col
    val elemEnd = col + node.nodeValue.length - 1

    var start = 0

    val newNodes = tokens
      .filter { t => elemStart <= t.location.start.col && t.location.end.col <= elemEnd }
      .flatMap { token =>
        val tokenStart = token.location.start.col - elemStart
        val tokenEnd = token.location.end.col - elemStart

        val beforeNodeOpt = if (tokenStart > start) {
          Some(dom.document.createTextNode(node.nodeValue.substring(start, tokenStart)))
        } else {
          None
        }

        val mainText = node.nodeValue.substring(tokenStart, tokenEnd + 1)
        val mainNode = makeUrlOpt(token)
          .map { url =>
            val anchor = dom.document.createElement("a")
            anchor.setAttribute("href", url)
            anchor.classList.add("codelab-link")
            anchor.textContent = mainText
            anchor.setAttribute("data-tippy-content", makeTooltipContent(token))
            Tippy.apply(anchor)
            anchor
          }
          .getOrElse(
            dom.document.createTextNode(mainText)
          )

        shouldLineBeHighlighted ||= selectedNodeIdOpt.exists { selectedNodeId =>
          token match {
            case u: Usage => u.definition.nodeId == selectedNodeId
            case d: Definition => d.nodeId == selectedNodeId
          }
        }

        start = tokenEnd + 1

        beforeNodeOpt.toList ++ Seq(mainNode)
      }

    val lastNodeOpt = if (start < node.nodeValue.length) {
      Some(dom.document.createTextNode(node.nodeValue.substring(start)))
    } else {
      None
    }

    Some(newNodes ++ lastNodeOpt.toList).filter(_.nonEmpty)
  }

  def walk(node: Node): Option[Seq[Node]] = {
    if (node.nodeType == Node.TEXT_NODE) {
      val newChildrenOpt = modify(node)
      col += node.nodeValue.length

      newChildrenOpt
    } else {
      val children = 0.until(node.childNodes.length).map { i => node.childNodes.item(i) }

      children.foreach { child =>
        walk(child).foreach { newChildren =>
          newChildren.foreach { newChild =>
            node.insertBefore(newChild, child)
          }
          node.removeChild(child)
        }
      }

      None
    }
  }

  def process(lineElem: Element): Boolean = {
    col = 1
    walk(lineElem)
    shouldLineBeHighlighted
  }
}
